neqo-common/src/socket.rs

```diff
@@ -0,0 +1,217 @@
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use std::{
+    io::{self, Error, ErrorKind, IoSlice, IoSliceMut},
+    net::{SocketAddr, UdpSocket},
+    os::fd::{AsRawFd, FromRawFd},
+};
+
+use nix::{
+    cmsg_space,
+    errno::Errno::{EAGAIN, EINTR},
+    sys::socket::{
+        recvmsg, sendmsg, setsockopt,
+        sockopt::{
+            IpDontFrag, IpRecvTos, IpRecvTtl, Ipv6DontFrag, Ipv6RecvHopLimit, Ipv6RecvTClass,
+        },
+        AddressFamily,
+        ControlMessage::{IpTos, IpTtl, Ipv6HopLimit, Ipv6TClass},
+        ControlMessageOwned, MsgFlags, SockaddrLike, SockaddrStorage,
+    },
+};
+
+use crate::Datagram;
+
+/// Binds a UDP socket to the specified local address.
+///
+/// # Arguments
+///
+/// * `local_addr` - The local address to bind the socket to.
+///
+/// # Returns
+///
+/// The bound UDP socket.
+///
+/// # Panics
+///
+/// Panics if the UDP socket fails to bind to the specified local address or if various
+/// socket options cannot be set.
+///
+/// # Notes
+///
+/// This function binds the UDP socket to the specified local address and performs additional
+/// configuration on the socket, such as setting socket options to request TOS and TTL
+/// information for incoming packets.
+#[allow(clippy::missing_errors_doc)]
+pub fn bind(local_addr: SocketAddr) -> io::Result<UdpSocket> {
+    let socket = match UdpSocket::bind(local_addr) {
+        Err(e) => {
+            panic!("Unable to bind UDP socket: {}", e);
+        }
+        Ok(s) => {
+            // Don't let the host stack or network path fragment our IP packets
+            // (RFC9000, Section 14).
+            let res = match local_addr {
+                SocketAddr::V4(..) => setsockopt(&s, IpDontFrag, &true),
+                SocketAddr::V6(..) => setsockopt(&s, Ipv6DontFrag, &true),
+            };
+            assert!(res.is_ok());
```

Comment:
```suggestion
            debug_assert!(res.is_ok());
```

If this fails, let's not worry too much.